py/sentry_streams/runner.py

```diff
@@ -32,56 +72,22 @@ def main() -> None:
         "topics": {
             "logical-events": "events",
             "transformed-events": "transformed-events",
-        }
+        },
+        "broker": "localhost:9092",
     }
 
-    # def recurse_edge(input_name: str, stream: Any) -> None:
-    #     for next_step_name in p.edges.get(input_name, ()):
-    #         print(f"Apply step: {next_step_name}")
-    #         next_step: WithInput = cast(WithInput, p.steps[next_step_name])
-    #         recurse_edge(
-    #             next_step_name, next_step.apply_edge(stream, environment_config)
-    #         )
-
-    # for source in p.sources:
-    #     print(f"Apply source: {source.name}")
-    #     env_source = source.apply_source(env, environment_config)
-    #     recurse_edge(source.name, env_source)
-
-    def iterate_edges(step_streams: dict[str, Any]) -> None:
-        while step_streams:
-            for input_name in list(step_streams):
-                output_steps = p.outgoing_edges[input_name]
-                input_stream = step_streams.pop(input_name)
-
-                if not output_steps:
-                    continue
-
-                # check if the inputs are fanning out
-                if len(output_steps) > 1:
-                    pass
-
-                else:
-                    output_step_name = output_steps.pop()
-
-                    # check if the inputs are fanning in
-                    if len(p.incoming_edges[output_step_name]) > 1:
-                        pass
-
-                    # 1:1 between input and output stream
-                    else:
-                        next_step: WithInput = cast(WithInput, p.steps[output_step_name])
-                        print(f"Apply step: {next_step.name}")
-                        output_stream = next_step.apply_edge(input_stream, environment_config)
-                        step_streams[next_step.name] = output_stream
+    pipeline: Pipeline = pipeline_globals["pipeline"]
+    # This will not be harcdoded in the future
+    runtime_config: StreamAdapter = FlinkAdapter(environment_config, env)
+    translator = RuntimeTranslator(runtime_config)
 
     step_streams = {}
 
-    for source in p.sources:
+    for source in pipeline.sources:
         print(f"Apply source: {source.name}")
-        env_source = source.apply_source(env, environment_config)
+        env_source = translator.translate_step(source)
         step_streams[source.name] = env_source
-        iterate_edges(step_streams)
+        iterate_edges(step_streams, pipeline, translator)
```

Comment:
Nit, what about passing the source to `iterate_edges` and let the function do the whole traversal rather than translate the source here?
If you pass the source and delegate to the `iterate_edges` method you do not have to initialize the mutable mapping `step_streams` here which makes it harder to understand the code.

Moreover the way in the current version you are translating steps in two places: one here and one inside `iterate_edges`. This adds cognitive overhead when trying to understand the algorithm.

Reply:
Also true. I think I did it this way to maintain the existing state of the code but there was no real reason for it to be structured this way.